python/src/iceberg/expression/literals.py

```diff
@@ -0,0 +1,502 @@
+
+    def __eq__(self, other):
+        return self.value == other.value
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __lt__(self, other):
+        return self.value < other.value
+
+    def __gt__(self, other):
+        return self.value > other.value
+
+    def __le__(self, other):
+        return self.value <= other.value
+
+    def __ge__(self, other):
+        return self.value >= other.value
+
+
+@singledispatch
+def literal(value) -> Literal:
+    """
+    A generic Literal factory to construct an iceberg Literal based on python primitive data type
+    using dynamic overloading
+
+    Args:
+        value(python primitive type): the value to be associated with literal
+
+    Example:
+        from iceberg.expressions.literals import literal
+        >>> literal(123)
+        IntegerLiteral(123)
+    """
+    raise TypeError(f"Unimplemented Type Literal for value: {str(value)}")
+
+
+@literal.register(bool)
+def _(value: bool) -> Literal[bool]:
+    return BooleanLiteral(value)
+
+
+@literal.register(int)
+def _(value: int) -> Literal[int]:
+    """
+    Upgrade to long if python int is outside the JAVA_MIN_INT and JAVA_MAX_INT
+    """
+    if value < IntegerType.min or value > IntegerType.max:
+        return LongLiteral(value)
+    return IntegerLiteral(value)
+
+
+@literal.register(float)
+def _(value: float) -> Literal[float]:
+    """
+    Upgrade to double if python float is outside the JAVA_MIN_FLOAT and JAVA_MAX_FLOAT
+    """
+    if value < FloatType.min or value > FloatType.max:
+        return DoubleLiteral(value)
+    return FloatLiteral(value)
+
+
+@literal.register(str)
+def _(value: str) -> Literal[str]:
+    return StringLiteral(value)
+
+
+@literal.register(uuid.UUID)
+def _(value: uuid.UUID) -> Literal[uuid.UUID]:
+    return UUIDLiteral(value)
+
+
+@literal.register(bytes)
+def _(value: bytes) -> Literal[bytes]:
+    return FixedLiteral(value)
+
+
+@literal.register(bytearray)
+def _(value: bytearray) -> Literal[bytes]:
+    return BinaryLiteral(value)
+
+
+@literal.register(Decimal)
+def _(value: Decimal) -> Literal[Decimal]:
+    return DecimalLiteral(value)
+
+
+class AboveMax(Literal[None], Singleton):
+    def __init__(self):
+        pass
+
+    def value(self):
+        raise ValueError("AboveMax has no value")
+
+    def to(self, type_var):
+        raise TypeError("Cannot change the type of AboveMax")
+
+    def __repr__(self):
+        return "AboveMax()"
+
+    def __str__(self):
+        return "AboveMax"
+
+
+class BelowMin(Literal[None], Singleton):
+    def __init__(self):
+        pass
+
+    def value(self):
+        raise ValueError("BelowMin has no value")
+
+    def to(self, type_var):
+        raise TypeError("Cannot change the type of BelowMin")
+
+    def __repr__(self):
+        return "BelowMin()"
+
+    def __str__(self):
+        return "BelowMin"
+
+
+class BooleanLiteral(Literal[bool]):
+    def __init__(self, value):
+        self._value = value
```

Comment:
Doesn't this need to call `super().__init__(value)` so that it goes through [the check for `None`](https://github.com/apache/iceberg/pull/4262/files#diff-22fa50c420ab8a210f0470d50d942b30397fe4052dab287a8795320bf9b709e1R64-R65)?

Reply:
Or maybe this should just not override `__init__`? Looks like that's what `LongLiteral` does.

Reply:
+1 I think removing all of the `__init__` methods and letting it fall back the super class is the way to go.

Reply:
Thank you, I removed `__init__` and it shall fallback to super